KSP integration fails on non-standard build types

Our project uses a non-standard build type on top of `debug` and `release`. When trying to add BDF and its KSP integration to the project, assembling that custom build type fails with the following exception:

```
./gradlew assembleStaging --stacktrace

org.gradle.api.internal.tasks.TaskDependencyResolveException: Could not determine the dependencies of task ':app:dexBuilderStaging'.
	at org.gradle.api.internal.tasks.CachingTaskDependencyResolveContext.getDependencies(CachingTaskDependencyResolveContext.java:70)
	at org.gradle.execution.plan.TaskDependencyResolver.resolveDependenciesFor(TaskDependencyResolver.java:49)
	(...)
Caused by: org.gradle.api.internal.tasks.DefaultTaskContainer$TaskCreationException: Could not create task ':app:compileStagingImplementations'.
	at org.gradle.api.internal.tasks.DefaultTaskContainer.taskCreationException(DefaultTaskContainer.java:721)
	(...)
Caused by: java.util.NoSuchElementException: Key staging is missing in the map.
	at kotlin.collections.MapsKt__MapWithDefaultKt.getOrImplicitDefaultNullable(MapWithDefault.kt:24)
	at kotlin.collections.MapsKt__MapsKt.getValue(Maps.kt:369)
	at app.cash.better.dynamic.features.BetterDynamicFeaturesPlugin$setupBaseCodegen$1$processTask$1.invoke(BetterDynamicFeaturesPlugin.kt:456)
	at app.cash.better.dynamic.features.BetterDynamicFeaturesPlugin$setupBaseCodegen$1$processTask$1.invoke(BetterDynamicFeaturesPlugin.kt:451)
	at app.cash.better.dynamic.features.BetterDynamicFeaturesPlugin$setupBaseCodegen$1.invoke$lambda$2(BetterDynamicFeaturesPlugin.kt:454)
	at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:66)
	at org.gradle.api.internal.DefaultMutationGuard$1.execute(DefaultMutationGuard.java:66)
	(...)
```

This is caused by [this overeager assumption in the Gradle plugin](https://github.com/cashapp/better-dynamic-features/blob/c3cf3cdf6b973f7ad84fcd45cf849b8d74727f21/gradle-plugin/src/main/kotlin/app/cash/better/dynamic/features/BetterDynamicFeaturesPlugin.kt#L410-L413), where only the standard build types are set up for KSP. Any custom build type would be missing here.

A more reactionary approach to set up the compile configuration from within the `onVariants` block could be a solution. I'm trying to find time shortly to address this myself.

### Version

`0.5.0-agp8.8.0`

### Repro steps

1. Clone https://github.com/mannodermaus/bdf-custom-buildtype-reproducer
2. `./gradlew assembleStaging`

--- a/gradle-plugin/src/main/kotlin/app/cash/better/dynamic/features/BetterDynamicFeaturesPlugin.kt
+++ b/gradle-plugin/src/main/kotlin/app/cash/better/dynamic/features/BetterDynamicFeaturesPlugin.kt
@@ -382,24 +382,23 @@ class BetterDynamicFeaturesPlugin : Plugin<Project> {
   }
 
   private fun Project.setupVariantCodegenDependencies(buildType: String): Configuration {
-    val compileConfiguration = configurations.create(CONFIGURATION_BDF_COMPILE_CLASSPATH(buildType)).apply {
-      isCanBeConsumed = false
-      isCanBeResolved = true
-      isVisible = false
+    val configurationName = CONFIGURATION_BDF_COMPILE_CLASSPATH(buildType)
+    val compileConfiguration = configurations.findByName(configurationName)
+      ?: run {
+        configurations.create(configurationName).apply {
+          isCanBeConsumed = false
+          isCanBeResolved = true
+          isVisible = false
 
-      attributes.apply {
-        attribute(BuildTypeAttr.ATTRIBUTE, project.objects.named(BuildTypeAttr::class.java, buildType))
-        attribute(ARTIFACT_TYPE, "android-classes-jar")
+          attributes.apply {
+            attribute(BuildTypeAttr.ATTRIBUTE, project.objects.named(BuildTypeAttr::class.java, buildType))
+            attribute(ARTIFACT_TYPE, "android-classes-jar")
+          }
+        }.also {
+          dependencies.add(configurationName, "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$KOTLIN_VERSION")
+          project.dependencies.add(configurationName, "app.cash.better.dynamic.features:runtime:$VERSION")
+        }
       }
-    }
-    dependencies.add(
-      CONFIGURATION_BDF_COMPILE_CLASSPATH(buildType),
-      "org.jetbrains.kotlin:kotlin-stdlib-jdk8:$KOTLIN_VERSION",
-    )
-    project.dependencies.add(
-      CONFIGURATION_BDF_COMPILE_CLASSPATH(buildType),
-      "app.cash.better.dynamic.features:runtime:$VERSION",
-    )
 
     return compileConfiguration
   }
@@ -407,11 +406,6 @@ class BetterDynamicFeaturesPlugin : Plugin<Project> {
   private fun Project.setupBaseCodegen(
     androidComponents: AndroidComponentsExtension<*, *, *>,
   ) {
-    val compileConfigurations = mapOf(
-      "debug" to setupVariantCodegenDependencies("debug"),
-      "release" to setupVariantCodegenDependencies("release"),
-    )
-
     val configuration = configurations.create(CONFIGURATION_BDF_IMPLEMENTATIONS).apply {
       isCanBeConsumed = false
       isCanBeResolved = true
@@ -453,7 +447,7 @@ class BetterDynamicFeaturesPlugin : Plugin<Project> {
         TypesafeImplementationsCompilationTask::class.java,
       ) { task ->
         task.generatedSources.set(implementationsTask.flatMap { it.generatedFilesDirectory })
-        val compileConfiguration = compileConfigurations.getValue(androidVariant.buildType!!)
+        val compileConfiguration = setupVariantCodegenDependencies(androidVariant.buildType!!)
         task.kotlinCompileClasspath.setFrom(project.provider { compileConfiguration.resolvedConfiguration.files })
       }
 

--- a/gradle-plugin/src/test/fixtures/custom-build-type/base/build.gradle
+++ b/gradle-plugin/src/test/fixtures/custom-build-type/base/build.gradle
@@ -1,0 +1,24 @@
+apply plugin: 'com.android.application'
+apply plugin: 'org.jetbrains.kotlin.android'
+apply plugin: 'com.google.devtools.ksp'
+apply plugin: 'app.cash.better.dynamic.features'
+
+android {
+  namespace "app.cash.better.dynamic.features.integration"
+
+  compileSdk 32
+  defaultConfig {
+    minSdk = 24
+
+    versionCode 1
+  }
+
+  buildTypes {
+    staging {
+      initWith debug
+      matchingFallbacks = ['debug']
+    }
+  }
+
+  dynamicFeatures = [':feature']
+}

--- a/gradle-plugin/src/test/fixtures/custom-build-type/base/src/main/AndroidManifest.xml
+++ b/gradle-plugin/src/test/fixtures/custom-build-type/base/src/main/AndroidManifest.xml
@@ -1,0 +1,1 @@
+<manifest />

--- a/gradle-plugin/src/test/fixtures/custom-build-type/build.gradle
+++ b/gradle-plugin/src/test/fixtures/custom-build-type/build.gradle
@@ -1,0 +1,13 @@
+buildscript {
+  apply from: "${projectDir.absolutePath}/../buildscript.gradle"
+}
+
+allprojects {
+  repositories {
+    maven {
+      url "file://${rootDir}/../../../../build/localMaven"
+    }
+    mavenCentral()
+    google()
+  }
+}

--- a/gradle-plugin/src/test/fixtures/custom-build-type/feature/build.gradle
+++ b/gradle-plugin/src/test/fixtures/custom-build-type/feature/build.gradle
@@ -1,0 +1,28 @@
+apply plugin: 'com.android.dynamic-feature'
+apply plugin: 'org.jetbrains.kotlin.android'
+apply plugin: 'com.google.devtools.ksp'
+apply plugin: 'app.cash.better.dynamic.features'
+
+android {
+  namespace "app.cash.better.dynamic.features.integration.feature"
+
+  compileSdk 32
+  defaultConfig {
+    minSdk = 24
+  }
+
+  buildTypes {
+    staging {
+      initWith debug
+      matchingFallbacks = ['debug']
+    }
+  }
+}
+
+dependencies {
+  implementation project(":base")
+}
+
+betterDynamicFeatures {
+  baseProject.set(project(":base"))
+}

--- a/gradle-plugin/src/test/fixtures/custom-build-type/feature/src/main/AndroidManifest.xml
+++ b/gradle-plugin/src/test/fixtures/custom-build-type/feature/src/main/AndroidManifest.xml
@@ -1,0 +1,14 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    xmlns:tools="http://schemas.android.com/tools"
+    xmlns:dist="http://schemas.android.com/apk/distribution">
+    <dist:module
+        dist:instant="false"
+        dist:title="Feature">
+        <dist:delivery>
+            <dist:install-time />
+        </dist:delivery>
+        <dist:fusing dist:include="true" />
+    </dist:module>
+</manifest>

--- a/gradle-plugin/src/test/fixtures/custom-build-type/gradle.properties
+++ b/gradle-plugin/src/test/fixtures/custom-build-type/gradle.properties
@@ -1,0 +1,2 @@
+org.gradle.jvmargs=-Xmx4g -XX:MaxMetaspaceSize=1g
+android.useAndroidX=true

--- a/gradle-plugin/src/test/fixtures/custom-build-type/settings.gradle
+++ b/gradle-plugin/src/test/fixtures/custom-build-type/settings.gradle
@@ -1,0 +1,4 @@
+apply from: "../settings.gradle"
+
+include ':base'
+include ':feature'

--- a/gradle-plugin/src/test/java/app/cash/better/dynamic/features/BetterDynamicFeaturesPluginTest.kt
+++ b/gradle-plugin/src/test/java/app/cash/better/dynamic/features/BetterDynamicFeaturesPluginTest.kt
@@ -850,6 +850,17 @@ class BetterDynamicFeaturesPluginTest {
     )
   }
 
+  @Test fun `custom build type is acknowledged`() {
+    val integrationRoot = File("src/test/fixtures/custom-build-type")
+
+    val gradleRunner = GradleRunner.create()
+      .withCommonConfiguration(integrationRoot)
+      .cleaned()
+      .withArguments(":base:assembleStaging", "--stacktrace")
+
+    gradleRunner.build()
+  }
+
   private fun clearLockfile(root: File) {
     root.lockfile().takeIf { it.exists() }?.delete()
   }

--- a/gradle-plugin/src/test/java/app/cash/better/dynamic/features/BetterDynamicFeaturesPluginTest.kt
+++ b/gradle-plugin/src/test/java/app/cash/better/dynamic/features/BetterDynamicFeaturesPluginTest.kt
@@ -856,9 +856,13 @@ class BetterDynamicFeaturesPluginTest {
     val gradleRunner = GradleRunner.create()
       .withCommonConfiguration(integrationRoot)
       .cleaned()
-      .withArguments(":base:assembleStaging", "--stacktrace")
+      .withArguments(":base:writeLockfile")
 
+    // Generate lockfile first, and then run assemble task
     gradleRunner.build()
+    val result = gradleRunner.withArguments(":base:assembleStaging", "--stacktrace").build()
+
+    assertThat(result.output).contains("BUILD SUCCESSFUL")
   }
 
   private fun clearLockfile(root: File) {

